runtime/doc/faq.txt

```diff
@@ -0,0 +1,684 @@
+<
+In case gdbserver puts the TUI as a background process, the TUI can become
+unable to read input from pty (and receives SIGTTIN signal) and/or output data
+(SIGTTOU signal). To force the TUI as the foreground process, you can add
+>
+    signal (SIGTTOU, SIG_IGN);
+    if (!tcsetpgrp(data->input.in_fd, getpid())) {
+        perror("tcsetpgrp failed");
+    }
+<
+to `tui.c:terminfo_start`.
+
+
+USING GDBSERVER IN TMUX ~
+
+Consider using a custom makefile
+https://github.com/neovim/neovim/wiki/Building-Neovim#custom-makefile to
+quickly start debugging sessions using the `gdbserver` method mentioned above.
+This example `local.mk` will create the debugging session when you type
+`make debug`.
+>make
+    .PHONY: dbg-start dbg-attach debug build
+
+    build:
+        @$(MAKE) nvim
+
+    dbg-start: build
+        @tmux new-window -n 'dbg-neovim' 'gdbserver :6666 ./build/bin/nvim -D'
+
+    dbg-attach:
+        @tmux new-window -n 'dbg-cgdb' 'cgdb -x gdb_start.sh ./build/bin/nvim'
+
+    debug: dbg-start dbg-attach
+<
+Here `gdb_start.sh` includes `gdb` commands to be called when the debugger
+starts. It needs to attach to the server started by the `dbg-start` rule. For
+example:
+>
+    target remote localhost:6666
+    br main
+<
+
+LOG FILE LOCATION ~
+
+Nvim's low-level logs are written to `$XDG_STATE_HOME/nvim/log` by default
+(typically `.local/state/nvim/log` on Posix platforms), but you can specify
+the location with the |$NVIM_LOG_FILE| environment variable. Debug builds
+write INFO-level messages to this log file. Non-debug builds only log
+ERROR-level messages.
+
+
+==============================================================================
+Design                                                         *faq-design*
+
+
+WHY NOT USE JSON FOR RPC? ~
+
+- JSON cannot easily/efficiently handle binary data
+- JSON specification is ambiguous: https://seriot.ch/parsing_json.php
+
+
+WHY EMBED LUA INSTEAD OF X? ~
+
+- Lua is a very small language, ideal for embedding. The biggest advantage of
+  Python/Ruby/etc is their huge collection of libraries, but that isn't
+  relevant for Nvim, where Nvim is the "batteries included" library:
+  introducing another stdlib would be redundant.
+- Lua 5.1 is a complete language: the syntax is frozen. This is great for
+  backwards compatibility.
+- Nvim also uses Lua internally as an alternative to C. Extra performance is
+  useful there, as opposed to a slow language like Python or Vim9script.
+- LuaJIT is one of the fastest runtimes on the planet, 10x faster than Python
+  and "Vim9script" https://vimhelp.org/vim9.txt.html, 100x faster than
+  Vimscript.
+- Python/JS cost more than Lua in terms of size and portability, and there are
+  already numerous Python/JS-based editors. So Python/JS would make Nvim
+  bigger and less portable, in exchange for a non-differentiating feature.
+
+See also:
+
+- Why Lua https://web.archive.org/web/20150219224654/https://blog.datamules.com/blog/2012/01/30/why-lua/
```

Comment:
the redis post is good, let's keep it

Reply:
Really? It doesn't really mention Lua in any detail. What's good about it?

Reply:
similar architecture as Nvim, similar strategy for batching work and avoiding a big RPC API